tests/primitives/test_retry.py

```diff
@@ -0,0 +1,148 @@
+from typing import Dict
+
+import httpx
+import pytest
+import tenacity
+from httpx._types import URLTypes
+from toloka.client.exceptions import ApiError
+from toloka.client.primitives.retry import SyncRetryingOverURLLibRetry
+from urllib3 import Retry
+
+
+class ExceptionRaisingRequester:
+    def __init__(self, exception):
+        self.exception = exception
+
+    def __call__(self, method: str, url: URLTypes, **kwargs):
+        raise self.exception
+
+
+@pytest.mark.parametrize(
+    'exception_to_raise', [
+        httpx.TimeoutException(''),
+        httpx.ConnectTimeout(''),
+        httpx.ReadTimeout(''),
+        httpx.WriteTimeout(''),
+        httpx.ConnectError(''),
+        httpx.ReadError(''),
+        httpx.WriteError(''),
+        httpx.CloseError(''),
+        httpx.ProtocolError(''),
+        httpx.LocalProtocolError(''),
+        httpx.RemoteProtocolError(''),
+    ]
+)
+def test_retrying_retries_relevant_exceptions(exception_to_raise):
+    retrying = SyncRetryingOverURLLibRetry(
+        base_url='http://example.com',
+        retry=Retry(backoff_factor=0, total=3),
+        reraise=True,
+    )
+    retrying_request = retrying.wraps(ExceptionRaisingRequester(exception_to_raise))
+
+    with pytest.raises(type(exception_to_raise)):
+        retrying_request('GET', 'http://example.com', timeout=0.1)
+    assert retrying.statistics['attempt_number'] == 4
+
+
+@pytest.mark.parametrize(
+    'exception_to_raise', [
+        httpx.NetworkError(''),  # not raised in httpcore, just a base class
+        httpx.PoolTimeout(''),
+        ValueError(),
+    ]
+)
+def test_retrying_does_not_retry_irrelevant_exceptions(exception_to_raise):
+    retrying = SyncRetryingOverURLLibRetry(
+        base_url='http://example.com',
+        retry=Retry(backoff_factor=0, total=3),
+        reraise=True,
+    )
+    retrying_request = retrying.wraps(ExceptionRaisingRequester(exception_to_raise))
+
+    with pytest.raises(type(exception_to_raise)):
+        retrying_request('GET', 'http://example.com', timeout=0.1)
+    assert retrying.statistics['attempt_number'] == 1
+
+
+@pytest.mark.timeout(1)
```

Comment:
Why do you use timeout?

Reply:
It is easy to break this test in a way so it will never stop